File: analysis_fw/reader.py
"""Read records from a .pb file.

Profile FW writes records inside a single wrapper message with a `repeated`
field (e.g. BlockDeviceStatLog { repeated BlockStatSample ... }). Protobuf's
repeated encoding delimits the records internally, so no external framing /
length prefix is used: each .pb file is one complete wrapper message.

We parse the whole file as the wrapper and iterate its repeated field. (If a
producer ever defines no wrapper, we fall back to treating the file as a single
record message.) A decode failure is turned into a clear, forwardable error.
"""
from __future__ import annotations

from pathlib import Path
from typing import Iterator

from google.protobuf.message import DecodeError

from .errors import InputError
from .registry import TableSchema


def iter_records(pb_path: Path, schema: TableSchema) -> Iterator:
    """Yield each record message from one .pb wrapper file.

    Reads the whole file into memory and parses it as one message — this is
    inherent to the wrapper format (the file is one message). Profile FW caps
    file size at rotation and splits big runs into several .pb files, so memory
    stays bounded per file.
    """
    data = pb_path.read_bytes()

    if schema.wrapper_cls is not None:
        wrapper = schema.wrapper_cls()
        try:
            wrapper.ParseFromString(data)
        except DecodeError as exc:
            raise InputError(_decode_error(pb_path, schema.wrapper_cls, data, exc)) from None
        yield from getattr(wrapper, schema.repeated_field)
        return

    # no wrapper defined: the whole file is a single record
    rec = schema.record_cls()
    try:
        rec.ParseFromString(data)
    except DecodeError as exc:
        raise InputError(_decode_error(pb_path, schema.record_cls, data, exc)) from None
    yield rec


def _decode_error(pb_path: Path, cls, data: bytes, exc: Exception) -> str:
    return (
        f"{pb_path.name}: could not parse the file as "
        f"{cls.DESCRIPTOR.full_name} ({exc}).\n"
        f"  file size: {len(data)} bytes; first bytes: {data[:32].hex(' ')}\n"
        f"  => the .pb likely does not match its .proto (wrong schema/version), "
        f"or the file is truncated/corrupt. Confirm the .proto with Profile FW."
    )

File: internal/reader/reader.go
// Package reader parses a .pb file and iterates its records.
//
// A .pb file IS one wrapper message holding a `repeated <record>` list —
// protobuf's repeated encoding delimits the records, so there is no framing or
// length prefix. A big run is split into several .pb files, each a complete
// wrapper.
package reader

import (
	"fmt"
	"os"

	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/types/dynamicpb"

	"github.com/vishal9359/Analysis-FW/internal/fwerr"
	"github.com/vishal9359/Analysis-FW/internal/registry"
)

// Records parses one .pb file and returns its records in order.
//
// A parse failure means the .pb does not match its .proto (schema drift) or is
// truncated — reported as an input error with enough detail to diagnose it.
func Records(path string, schema *registry.TableSchema) (protoreflect.List, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fwerr.InputWrap(err, "cannot read %s", path)
	}

	if schema.Wrapper == nil {
		return nil, fwerr.Schema("%s: no wrapper message (a message with a repeated "+
			"record field) found in the .proto", path)
	}

	wrapper := dynamicpb.NewMessage(schema.Wrapper)
	if err := proto.Unmarshal(data, wrapper); err != nil {
		return nil, fwerr.InputWrap(err,
			"%s: could not parse as %s — the .pb does not match its .proto "+
				"(schema drift) or is truncated/corrupt. First bytes: %s",
			path, schema.Wrapper.Name(), firstBytes(data))
	}
	return wrapper.Get(schema.RepeatedField).List(), nil
}

func firstBytes(b []byte) string {
	n := 16
	if len(b) < n {
		n = len(b)
	}
	return fmt.Sprintf("% x", b[:n])
}
